### PR TITLE
KFRD-9 MessageQueue was implemented

### PR DESCRIPTION
**Pull request by MedvedevBW**

### CODE REVIEW

**Comment by MedvedevBW** [comment]
@placccebo

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+*/
+package com.bwsw.kafka.reader
+
+import com.bwsw.kafka.reader.entities.InputEnvelope
+
+import scala.collection.mutable.ListBuffer
+import scala.collection.JavaConverters._
+
+/**
+  * Class is responsible for retrieving ConsumerRecords from Consumer
```
Comment (placccebo) [inline src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: Class is an intermediate buffer of Kafka events used to decrease a count of requests to Kafka if there is a need to retrieve a small number of messages, e.g. one. Class retrieves a list of ConsumerRecords and convert each of them to InputEnvelope

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  * and keeping it as InputEnvelope entities after its conversion
+  *
+  */
+class MessageQueue[K,V](consumer: Consumer[K,V]) {
+  private var buffer = new ListBuffer[InputEnvelope[V]]
+
+  /**
+    * Retrieves 'n' InputEnvelope from the buffer,
```
Comment (placccebo) [inline src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala:39]: Retrieves 'n' InputEnvelope from the buffer, if buffer does not have enough entities, it tries to retrieve the available data from the consumer

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      buffer.take(sizeAfterFill)
+    } else {
+      buffer.take(n)
+    }
+
+    buffer.remove(0, envelopes.size)
+    envelopes.toList
+  }
+
+  private def fill(): Unit = {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: add description:
Retrieves a list of ConsumerRecords, convert each of them to InputEnvelope and put to a buffer

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    * @param n count of InputEnvelope entities to extract
+    */
+  def take(n: Int): List[InputEnvelope[V]] = {
+    if (buffer.size < n) {
+      fill()
+    }
+
+    val sizeAfterFill = buffer.size
+
+    val envelopes = if (sizeAfterFill <= n) {
```
Comment (placccebo) [inline src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: maybe "<" would be better than "<=" ?

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+      case Success(x) =>
+        Try(testConsumer.close())
+        x
+      case Failure(e: Throwable) =>
+        testConsumer.close()
+        throw e
+    }
+  }
+
+  "fill" should "put InputEnvelope list to buffer" in { fixture =>
```
Comment (placccebo) [inline src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: change a test description according to proposed (above) "fill()" method description

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    val buffer = messageClass.getDeclaredField("buffer")
+    buffer.setAccessible(true)
+
+    val actualInputEnvelopes = buffer.get(fixture.messageQueue).asInstanceOf[ListBuffer[InputEnvelope[String]]].toList
+
+    assert(actualInputEnvelopes == fixture.expectedInputEnvelopes)
+  }
+
+  "take" should "get N InputEnvelopes from buffer if buffer has enough entities" in { fixture =>
```
Comment (placccebo) [inline src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: N -> specified number of 

get or return?

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  "fill" should "put InputEnvelope list to buffer" in { fixture =>
+    val messageClass = classOf[MessageQueue[String,String]]
+
+    val fill = messageClass.getDeclaredMethod("fill")
+    fill.setAccessible(true)
+    fill.invoke(fixture.messageQueue)
+
+    val buffer = messageClass.getDeclaredField("buffer")
```
Comment (placccebo) [inline src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: this is unnecessary

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+  "take" should "get N InputEnvelopes from buffer if buffer has enough entities after execution method fill()" in { fixture =>
+    val size = fixture.expectedInputEnvelopes.size
+    fixture.buffer.set(fixture.messageQueue, new ListBuffer[InputEnvelope[String]])
+
+    assert(fixture.expectedInputEnvelopes.take(size - 1) == fixture.messageQueue.take(size - 1))
+    assert(fixture.buffer.get(fixture.messageQueue).asInstanceOf[ListBuffer[String]].size == 1)
+  }
+
+  "take" should "get all InputEnvelopes from buffer if buffer does not have enough entities after execution method fill()" in { fixture =>
```
Comment (placccebo) [inline src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: all available InputEnvelopes

**Review by placccebo — CHANGES_REQUESTED** [review]

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala (hunk 4 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: I think that actually it's not so important. Fixed

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala (hunk 3 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: ok

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala (hunk 7 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: removed

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala (hunk 5 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: ok

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala (hunk 6 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: return ... which retrieves ...

File: src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala (hunk 8 above)
Comment (MedvedevBW) [reply src/test/scala/com/bwsw/kafka/reader/MessageQueueTestSuite.scala]: ok

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala (hunk 2 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala:39]: ok

File: src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala (hunk 1 above)
Comment (MedvedevBW) [reply src/main/scala/com/bwsw/kafka/reader/MessageQueue.scala]: ok

**Review by placccebo — APPROVED** [review]

**Pull request merged**